services/blockchain-indexer/shared/utils/validatorUtils.js

```diff
@@ -51,7 +68,36 @@ const getAddressByName = async (name) => {
 	return null;
 };
 
+const calcCommission = async (generatorAddress, reward) => {
+	const commissionsTable = await getCommissionsTable();
+	const [{ commission: currentCommission }] = await commissionsTable
+		.find({ address: generatorAddress, sort: 'height:desc', limit: 1 }, 'commission');
+	const commission = (BigInt(reward) * BigInt(currentCommission)) / MAX_COMMISSION;
+	return commission;
+};
+
+const calcSelfStakeReward = async (generatorAddress, reward, commission) => {
+	let selfStakeReward = BigInt('0');
+
+	const stakesTable = await getStakesTable();
+	const stakerInfo = await stakesTable.find(
+		{ validatorAddress: generatorAddress }, ['stakerAddress', 'amount'],
+	);
+
+	if (stakerInfo.length) {
+		const selfStakesInfo = stakerInfo.filter(stake => stake.stakerAddress === generatorAddress);
+		const selfStakes = selfStakesInfo.reduce((a, b) => BigInt(a.amount) + BigInt(b.amount));
```

Comment:
I think this information could be get directly from some endpoint instead of being re-calculated here. 

In the PoS LIP we have the `getDelegate(address: Address)` endpoint (to be renamed `getValidator(address)` after terminology change) which returns all the information of the validator in the state store. 

The resulting object follows [this schema](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0057.md#json-schema-1) and contains `selfVotes` (to be renamed `selfStake`) and `totalVotesReceived` (to be renamed `totalStake`), so I guess you can get this values from there instead of scanning the whole stake database and adding the values, which increases the running time in the UI

Reply:
Assuming the current blockchain height is at height `hCurrent` and the highlighted calculation is made at height `h` which is lower than `hCurrent`. The issue with querying `getValdiator` is that the information that the node retrieves will be the updated state as of height `hCurrent` and not at height `h`. Hence, to compute the `totalSelfStake` and `totalCommission` values for every validator, Lisk Service requires the `totalStake` and `selfStake` values for the block generator at height `h` (lesser than equal to `hCurrent`). And, to calculate the `totalStake` and `selfStake` values at height `h`, we will need to be compute the `totalStake` and filter the `selfStake` values for the block generator at height `h`.

We can discuss this offline in detail, if the above comment is not clear enough.

Reply:
@gkoumout Assuming the current blockchain height is at height `hCurrent` and the highlighted calculation is made at height `h` which is lower than `hCurrent`. The issue with querying `getValdiator` is that the information that the node retrieves will be the updated state as of height `hCurrent` and not at height `h`. Hence, to compute the `totalSelfStake` and `totalCommission` values for every validator, Lisk Service requires the `totalStake` and `selfStake` values for the block generator at height `h` (lesser than equal to `hCurrent`). And, to calculate the `totalStake` and `selfStake` values at height `h`, we will need to be compute the `totalStake` and filter the `selfStake` values for the block generator at height `h`.

We can discuss this offline in detail, if the above comment is not clear enough.

Reply:
After discussing with Sameer and getting a better understanding of the explanation, can be resolved from my side.